File: EmarsysMobileShop/EmarsysMobileShop/CartViewController.m
//
//  CartViewController.m
//  EmarsysMobileShop
//

@import Foundation;

#import "CartViewController.h"
#import "AppDelegate.h"
#import "ItemDetailViewController.h"
#import "Cart+EmarsysMobileExtensions.h"
#import "DataSource.h"
#import "RecommendedItemsCell.h"
#import "RecommendedItem.h"
#import "CartItemCell.h"

@implementation CartViewController

static NSString *const reuseIdentifier = @"ItemCell";

- (void)viewDidLoad {
    [super viewDidLoad];

    [[NSNotificationCenter defaultCenter] addObserver:self.cartItemsTableView
                                             selector:@selector(reloadData)
                                                 name:@"CartChanged"
                                               object:nil];

    self.recommendResults = [[NSMutableArray alloc] init];
    self.cartItemsTableView.dataSource = self;
    self.cartItemsTableView.delegate = self;
    self.recommendedCollectionView.dataSource = self;
    self.recommendedCollectionView.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    // Deselect all rows
    for (NSIndexPath *indexPath in _cartItemsTableView
             .indexPathsForSelectedRows) {
        [self.cartItemsTableView deselectRowAtIndexPath:indexPath animated:NO];
    }

    [self sendRecommend];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    if ([Cart sharedCart].cartItems.count == 0) {
        self.buyButton.enabled = false;
    } else {
        self.buyButton.enabled = true;
    }
    return [Cart sharedCart].cartItems.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    CartItemCell *cell =
        [tableView dequeueReusableCellWithIdentifier:@"CartItemCell"];
    NSAssert(cell, @"No cell");

    CartItem *cartItem = [Cart sharedCart].cartItems[indexPath.row];

    [cell.cost
        setText:[NSString stringWithFormat:@"$%.02f", cartItem.item.price *
                                                          cartItem.count]];
    [cell.title setText:cartItem.item.title];
    [cell.count setText:[NSString stringWithFormat:@"%d", cartItem.count]];

    NSURL *url = [NSURL URLWithString:cartItem.item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {

          if (!data) {
              return;
          }

          UIImage *image = [UIImage imageWithData:data];
          if (!image) {
              return;
          }

          dispatch_async(dispatch_get_main_queue(), ^{
            CartItemCell *updateCell =
                (id)[tableView cellForRowAtIndexPath:indexPath];
            if (!updateCell) {
                return;
            }
            updateCell.image.image = image;
          });

        }];
    [task resume];

    return cell;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"CartShowItem"]) {
        ItemDetailViewController *ctr = segue.destinationViewController;
        ctr.item =
            [Cart sharedCart]
                .cartItems[self.cartItemsTableView.indexPathForSelectedRow.row]
                .item;
    } else if ([segue.identifier isEqualToString:@"CartShowRecommendedItem"]) {
        ItemDetailViewController *ctr = segue.destinationViewController;
        RecommendedItem *recItem = sender;
        ctr.item = recItem.item;
    }
}

- (BOOL)tableView:(UITableView *)tableView
    canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView
    commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
     forRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView beginUpdates];
    [[Cart sharedCart].cartItems removeObjectAtIndex:indexPath.row];
    [tableView deleteRowsAtIndexPaths:@[ indexPath ]
                     withRowAnimation:UITableViewRowAnimationAutomatic];
    [tableView endUpdates];
    [self sendRecommend];
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:@"Remove from cart was successful."
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (IBAction)buy:(id)sender {
    EMTransaction *transaction = [[EMTransaction alloc] init];
    /// Report the list of items in the visitor's shopping cart. This command
    /// should be
    /// called on every page. Make sure all cart items are passed to the
    /// command. If
    /// the visitor's cart is empty, send the empty array [].
    NSArray<EMCartItem *> *cartItems = [[Cart sharedCart] convertItems];
    [transaction setCart:cartItems];

    /// Report a purchase event. This command should be called on the order
    /// confirmation page to report successful purchases. Also make sure all
    /// purchased items are passed to the command.
    NSString *orderID = [[NSUUID UUID] UUIDString];
    [transaction setPurchase:orderID ofItems:cartItems];

    /// Send transaction to the recommender server.
    EMSession *emarsysSession = [EMSession sharedSession];
    [emarsysSession sendTransaction:transaction
                       errorHandler:^(NSError *_Nonnull error) {
                         NSLog(@"Error: %@ %@", error, [error userInfo]);
                       }];

    [[Cart sharedCart].cartItems removeAllObjects];
    [self.cartItemsTableView reloadData];
    [self sendRecommend];

    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:@"Buy was successful."
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

- (NSInteger)numberOfSectionsInCollectionView:
    (UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section {
    return self.recommendResults.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    RecommendedItem *cell =
        [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier
                                                  forIndexPath:indexPath];

    // Configure the cell

    NSAssert(cell, @"No cell");

    Item *item = self.recommendResults[indexPath.row];

    [cell.title setText:item.title];
    cell.item = item;
    NSURL *url = [NSURL URLWithString:item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {

          if (!data) {
              return;
          }

          UIImage *image = [UIImage imageWithData:data];

          if (!image) {
              return;
          }

          dispatch_async(dispatch_get_main_queue(), ^{
            RecommendedItem *updateCell =
                (id)[collectionView cellForItemAtIndexPath:indexPath];

            if (!updateCell) {
                return;
            }

            [updateCell.image setImage:image];
          });
        }];
    [task resume];

    return cell;
}

- (void)sendRecommend {

    EMSession *emarsysSession = [EMSession sharedSession];
    EMTransaction *transaction = [[EMTransaction alloc] init];

    /// Report the list of items in the visitor's shopping cart. This command
    /// should be
    /// called on every page. Make sure all cart items are passed to the
    /// command. If
    /// the visitor's cart is empty, send the empty array [].
    NSArray<EMCartItem *> *cartItems = [[Cart sharedCart] convertItems];
    [transaction setCart:cartItems];

    self.recommendResults = [[NSMutableArray alloc] init];

    EMRecommendationRequest *recommend =
        [[EMRecommendationRequest alloc] initWithLogic:@"CART"];
    recommend.limit = 10;
    recommend.completionHandler = ^(EMRecommendationResult *_Nonnull result) {

      // Process result
      NSLog(@"%@", result.featureID);
      [result.products
          enumerateObjectsUsingBlock:^(EMRecommendationItem *_Nonnull obj,
                                       NSUInteger idx, BOOL *_Nonnull stop) {

            Item *item = [[Item alloc] initWithItem:obj];
            [self.recommendResults addObject:item];
          }];

      dispatch_async(dispatch_get_main_queue(), ^{
        [self.recommendedCollectionView reloadData];
      });
    };
    [transaction recommend:recommend];

    // Firing the transaction. Should be the last call on the page, called only
    // once.
    [emarsysSession sendTransaction:transaction
                       errorHandler:^(NSError *_Nonnull error) {
                         // Some error happened
                         NSLog(@"value: %@", error);
                         return;
                       }];
}

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/CartViewController.swift
//
//  CartViewController.swift
//  EmarsysMobileShopSwift
//
//

import UIKit
import EmarsysMobile

class CartViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, UICollectionViewDelegate, UICollectionViewDataSource {
    
    @IBOutlet weak var buyButton: UIBarButtonItem!
    @IBOutlet weak var cartItemsTableView: UITableView!
    @IBOutlet weak var recommendedCollectionView: UICollectionView!
    
    var recommendResults = [Item]()
    
    @IBAction func buy(sender: UIBarButtonItem) {
        let transaction = EMTransaction()
        let cartItems = Cart.sharedCart.convertItems()
        transaction.setCart(cartItems)
        
        let orderID = NSUUID().UUIDString
        transaction.setPurchase(orderID, ofItems: cartItems)
        
        let emarsysSession = EMSession.sharedSession()
        emarsysSession.sendTransaction(transaction, errorHandler: {(error) in
            // Some error happened
            NSLog("value: %@", error)
            return
        })
        
        Cart.sharedCart.cartItems.removeAll()
        cartItemsTableView .reloadData()
        sendRecommend()
        
        let alert = UIAlertController(title: "Buy was successful.", message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        NSNotificationCenter.defaultCenter().addObserver(cartItemsTableView, selector: "reloadData", name: "CartChanged", object: nil)
        
        recommendResults = [Item]()
        cartItemsTableView.dataSource = self
        cartItemsTableView.delegate = self
        recommendedCollectionView.dataSource = self
        recommendedCollectionView.delegate = self
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        sendRecommend()
    }
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let count = Cart.sharedCart.cartItems.count
        buyButton.enabled = count != 0
        return count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("CartItemCell") as! CartItemCell
        
        if indexPath.row < Cart.sharedCart.cartItems.count {
            let cartItem = Cart.sharedCart.cartItems[indexPath.row]
            
            cell.title.text = cartItem.item.title
            cell.count.text = String(format: "%d", cartItem.count)
            
            let url = NSURL(string: cartItem.item.image!)
            let session = NSURLSession.sharedSession()
            
            session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
                guard let data = data else { return }
                guard let image = UIImage(data: data) else { return }
                dispatch_async(dispatch_get_main_queue()) {
                    if let updateCell = tableView.cellForRowAtIndexPath(indexPath) {
                        let cartItemCell = updateCell as! CartItemCell
                        cartItemCell.itemImage.image = image
                    }
                }
            }).resume()
        }
        
        return cell
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "CartShowItem" {
            let ctr = segue.destinationViewController as! ItemDetailViewController
            ctr.item = Cart.sharedCart.cartItems[cartItemsTableView.indexPathForSelectedRow!.row].item
        } else if segue.identifier == "CartShowRecommendedItem" {
            let ctr = segue.destinationViewController as! ItemDetailViewController
            let recommendedItem = sender as! RecommendedItem
            ctr.item = recommendedItem.item
        }
    }
    
    func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        return true
    }
    
    func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        tableView.beginUpdates()
        Cart.sharedCart.cartItems.removeAtIndex(indexPath.row)
        tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimation.Automatic)
        tableView.endUpdates()
        
        sendRecommend()
        
        let alert = UIAlertController(title: "Remove from cart was successful.", message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return recommendResults.count
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("ItemCell", forIndexPath: indexPath) as! RecommendedItem
        let item = recommendResults[indexPath.row]
        cell.title.text = item.title
        cell.item = item
        
        let url = NSURL(string: item.image!)
        let session = NSURLSession.sharedSession()
        session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            dispatch_async(dispatch_get_main_queue()) {
                guard let cell = collectionView.cellForItemAtIndexPath(indexPath) else {
                    NSLog("Failed to get cell for index path %@", indexPath)
                    return
                }
                let updateCell = cell as! RecommendedItem
                updateCell.image.image = image
            }
            
            
        }).resume()
        
        return cell
    }
    
    func sendRecommend() {
        let emarsysSession = EMSession.sharedSession()
        let transaction = EMTransaction()
        let cartItems = Cart.sharedCart.convertItems()
        transaction.setCart(cartItems)
        
        let recommend = EMRecommendationRequest(logic: "CART")
        
        recommend.limit = 10
        recommend.completionHandler = {result in
            // Process result
            for obj in result.products {
                let item = Item(withItem: obj)
                self.recommendResults.append(item)
            }
            
            dispatch_async(dispatch_get_main_queue()) {
                self.recommendedCollectionView.reloadData()
            }
        }
        
        transaction.recommend(recommend)
        
        // Firing the transaction. Should be the last call on the page, called only once.
        emarsysSession.sendTransaction(transaction, errorHandler: {(error) in
            // Some error happened
            NSLog("value: %@", error)
            return
        })
    }
}
